### ISSUE TITLE
void-returning foreign functions given unit type

### ISSUE THREAD

**Issue by DrBearhands** [idris-lang/Idris2#1843]
Came across what is probably a niche problem. In the documentation, foreign imperative functions with no return value are declared as returning a unit. E.g. this example:

```Idris
%foreign "browser:lambda:x=>{document.body.innerHTML = x}"
prim__setBodyInnerHTML : String -> PrimIO ()
```

taken from: https://idris2.readthedocs.io/en/latest/backends/javascript.html

In the javascript backend at least, this function does not actually return a unit, but `undefined`. This can result in runtime exceptions e.g. if the fake unit is pattern-matched. On the other hand, while incorrect, pretending the foreign function returns unit (without adding the overhead of generating a real unit value) seems useful in the majority of situations.

I suspect this problem will resolve itself with performance improvements to the backend, as a switch on unit seems redundant, but it might be a good idea to mention the problem in the docs.

**Comment by stefan-hoeck** [comment]
Interesting. It would be pretty straight forward (in the JS backend at least) to identify Idris2 `()` with JS `undefined` in pattern matches and constructor applications. Do you think @DrBearhands this would resolve this issue or can you see other problems with such a "fix"?

**Comment by stefan-hoeck** [comment]
So, here is a pathological example. We expect the following program to print
```
a test
foo
a test
```
but it actually prints
```
a test
foo
oopsie
```

```idris
test : () -> IO ()
test () = putStrLn "a test"
test _  = putStrLn "oopsie"

main : IO ()
main = do test ()
          putStrLn "foo" >>= test
```

The pattern match on `()` is converted to a switch on the integer constant `0`, so it might not be trivial to figure out that this actually came for `()` in the first place. If, however, the catch all case is removed from `test`, the JS backend optimizes away the single case switch statement. So, at least in the JS statement, this seems not to be a problem unless one writes nonsensical pattern matches as shown above.

I wonder, however, why Idris does not complain about an unreachable case in the code above...

**Comment by DrBearhands** [comment]
> Do you think @DrBearhands this would resolve this issue or can you see other problems with such a "fix"?

I suspect it would work, but frankly I don't know Idris well enough to make that call. I'd be concerned about weird behavior of undefined and runtime type information. E.g. for the former `undefined >= undefined` is false, but `undefined === undefined` is true. 

Concerning your pathological example, I didn't get that optimization, so I thought it didn't exist. Will have to investigate and return with a minimal example.

**Comment by Z-snails** [comment]
I don't think it would be hard to detect unit, then optimise away the entire case statement. I'll do that just as soon as I remember where we do all those optimisations (one of which I wrote 😂)

**Comment by stefan-hoeck** [comment]
> I don't think it would be hard to detect unit, then optimise away the entire case statement. I'll do that just as soon as I remember where we do all those optimisations (one of which I wrote joy)

The JS backend itself has code that optimizes away single case (or single default case) pattern matches in the `ToAst` module. At that point, however, unit might already have been converted to an integer constant (0), so it might no longer be possible to fix the pathological example mentioned above.

@Z-snails: If optimizing single case pattern matches already happens at an earlier stage, it might make sense to reflect this in the data constructors of the IRs so that backend writers can be sure that they are dealing with at least two different cases, no?

**Comment by Z-snails** [comment]
In the case of single constructor data-types with more than 0 arguments, case is the only way (in CExp) to access those arguments, so unit-like types are the only ones you can safely remove.
I think it would make sense to have a `List1` of alts, however I don't think it's really that useful, as it's easy to spot for backend authors, and it would be another breaking change.

**Comment by DrBearhands** [comment]
I think there might be 2 issues at hand here: (1) optimization is missed, and (2) bad values are used for a type when following the FFI examples in the documentation. I fear that optimizing away case statements for units is only fixing a symptom, and may result in more obscure errors elsewhere - though admittedly I cannot think of any - or cause problems with future changes.

**Comment by gallais** [comment]
> I wonder, however, why Idris does not complain about an unreachable case in the code above...

Could be something like #550

**Comment by stefan-hoeck** [comment]
> I think there might be 2 issues at hand here: (1) optimization is missed, and (2) bad values are used for a type when following the FFI examples in the documentation. I fear that optimizing away case statements for units is only fixing a symptom, and may result in more obscure errors elsewhere - though admittedly I cannot think of any - or cause problems with future changes.

I agree, and I actually think that converting `()` to `undefined` in the JS backend is the solution to this (we should of course mention this in the docs). To the best of my knowledge, implementations of Idris functions and interfaces involving unit do not go via the JS FFI (`Eq` and `Ord` for instance are implemented using pattern matching), so strange behavior as mentioned for instance for the  JS `>=` operator will not affect us here.

By the way, this is not limited to the JS backend. The example I posted behaves the same in the default Chez Scheme backend. The fix @Z-snails is working on will of course affect all backends, but the underlying inaccuracy and potential source of bugs as described by @DrBearhands remains.

**Comment by stefan-hoeck** [comment]
Just checked the example against the refc-backend: It segfaults.

**Comment by buzden** [comment]
> > I wonder, however, why Idris does not complain about an unreachable case in the code above...
> 
> Could be something like #550

More like #1079
